lib/hooks/userspace/hook-grpc.js

```diff
@@ -365,43 +365,79 @@ function wrapBidi(namespace, handlerSet, requestName) {
-  return function registerTrace(name, handler, serialize, deserialize, method_type) {
-    // register(n, h, s, d, m) is called in addService once for each service method.
-    // Its role is to assign the serialize, deserialize, and user logic handlers
-    // for each exposed service method. Here, we wrap these functions depending on the
-    // method type.
-    var namespace = cls.getNamespace();
-    if (!namespace) {
-      agent.logger.info('gRPC: no namespace found, ignoring request');
-      return register.apply(this, arguments);
-    }
-    var result = register.apply(this, arguments);
-    var handlerSet = this.handlers[name];
-    var requestName = 'grpc:' + name;
-    // Proceed to wrap methods that are invoked when a gRPC service call is made.
-    // In every case, the function 'func' is the user-implemented handling function.
-    if (method_type === 'unary') {
-      wrapUnary(namespace, handlerSet, requestName);
-    } else if (method_type === 'server_stream') {
-      wrapServerStream(namespace, handlerSet, requestName);
-    } else if (method_type === 'client_stream') {
-      wrapClientStream(namespace, handlerSet, requestName);
-    } else if (method_type === 'bidi') {
-      wrapBidi(namespace, handlerSet, requestName);
-    } else {
-      agent.logger.warn('gRPC Server: Unrecognized method_type ' + method_type);
-    }
-    return result;
+function wrapServer(server) {
+  var unwrapFnList = [];
+  function serverRegisterWrap(original) {
+    return function registerTrace(name, handler, serialize, deserialize, method_type) {
+      // register(n, h, s, d, m) is called in addService once for each service
+      // method. Its role is to assign the serialize, deserialize, and user
+      // logic handlers for each exposed service method. Here, we wrap these
+      // functions depending on the method type.
+      var namespace = cls.getNamespace();
+      if (!namespace) {
+        agent.logger.info('gRPC: no namespace found, ignoring request');
+        return original.apply(this, arguments);
+      }
+      var result = original.apply(this, arguments);
+      var handlerSet = this.handlers[name];
+      var requestName = 'grpc:' + name;
+      // Proceed to wrap methods that are invoked when a gRPC service call is
+      // made. In every case, the function 'func' is the user-implemented
+      // handling function.
+      if (method_type === 'unary') {
+        wrapUnary(namespace, handlerSet, requestName);
+      } else if (method_type === 'server_stream') {
+        wrapServerStream(namespace, handlerSet, requestName);
+      } else if (method_type === 'client_stream') {
+        wrapClientStream(namespace, handlerSet, requestName);
+      } else if (method_type === 'bidi') {
+        wrapBidi(namespace, handlerSet, requestName);
+      } else {
+        agent.logger.warn('gRPC Server: Unrecognized method_type ' +
+          method_type);
+      }
+      unwrapFnList.push(function() {
```

Comment:
I believe storing these functions will introduce a memory leak as they will hold reference to pieces of the server after the application code no longer has need of them. I don't think this is as big a problem as if it were a client library patch since servers shouldn't be created/destroyed frequently. @ofrobots Is this a leak we should be concerned about?